core/src/main/java/org/apache/carbondata/core/scan/expression/RangeExpressionEvaluator.java

```diff
@@ -223,12 +223,27 @@ private void fillExpressionMap(Map<String, List<FilterModificationNode>> filterE
       addFilterExpressionMap(filterExpressionMap, currentNode, parentNode);
     }
 
-    for (Expression exp : currentNode.getChildren()) {
-      if (null != exp) {
-        fillExpressionMap(filterExpressionMap, exp, currentNode);
-        if (exp instanceof OrExpression) {
-          replaceWithRangeExpression(filterExpressionMap);
-          filterExpressionMap.clear();
+
+    // In case of Or Exp we have to evaluate both the subtrees of expression separately
+    // else it will combine the results of both the subtrees into one expression
+    // which wont give us correct result
+    if (currentNode instanceof OrExpression) {
+      Map<String, List<FilterModificationNode>> filterExpressionMapNew =
+          new HashMap<>(CarbonCommonConstants.DEFAULT_COLLECTION_SIZE);
+      if (null != ((OrExpression) currentNode).left) {
```

Comment:
Don't typecast multiple time, please do it once

Reply:
Done.